### DISCUSSION TITLE
Rejoin into network .

### DISCUSSION THREAD

**Discussion by mithunGhosh13** [openthread/openthread#7607] [Q&A] [answered]
How could i get reconnected into the same network while my device got disconnected due to some reason?

**Comment by jwhui** [comment, marked as answer]
Thread devices should automatically look to reconnect to an existing network.

A Thread end device will periodically look for a parent to attach to while it is detached.

A Thread router-capable device will form its own partition and become a leader if it cannot discover any other routers around. However, a router-capable device does discover another neighboring partition, the devices should attempt to merge the partitions.

**Reply by mithunGhosh13** [reply]
is any network data needs to be store while it in paired? if so then what are the parameters needs to be store?

**Reply by jwhui** [reply]
If you follow the [Operational Dataset Quick Start Guide](https://github.com/openthread/openthread/blob/main/src/cli/README_DATASET.md#quick-start), OpenThread will handle non-volatile storage of the Operational Dataset.

**Reply by mithunGhosh13** [reply]
The scenario is like let , Device is in paired and communication is running fine, if power goes  the device got disconnected and it should connect automatically into the previously paired network when it's powered again. for that what are way i can flow to rejoin?

**Reply by jwhui** [reply]
If you are using the OT CLI example app, it does not automatically start Thread after boot. You can modify the code to call `otIp6IsEnabled()` and `otThreadSetEnabled()`.

**Reply by mithunGhosh13** [reply]
Actually i am not configuring network from boot. Thing is like after getting response from parents we needs to press key then only i start the thread state machine. where the steps is like 
1. doing communication init ( otSysInit(),platform_init(),otInstanceInitSingle(), otPlatFlashErase(),otThreadSetLinkMode(),otIp6SetEnabled(),DataBase_Init() and KNX initialization  ) these are the steps am flowing for init)
2. then calling otJoinerStart(). for join ing.
3. then otThreadSetEnabled() for pair.
4. then after got paired am satying into sleep state and comes out when wake up time elapse. and the step 4 process continuing.

**Reply by mithunGhosh13** [reply]
no i am not using the OT CLI example app

**Reply by jwhui** [reply]
When the device successfully attaches to the network for the first time, it will retrieve the Active Operational Dataset from its parent and store it in non-volatile memory.

**Reply by mithunGhosh13** [reply]
ok, Then after wats needs to do?

**Reply by jwhui** [reply]
Just make sure to enable Thread by calling `otIp6IsEnabled()` and `otThreadSetEnabled()`.

**Reply by mithunGhosh13** [reply]
i can see using otDatasetGetActive() we can able to see the active network data sets right and we need to store that data sets into NVM right? after that when power comes is it need to set those data sets using otDatasetSetActive() after reading data set from NVM. Then we need to call otIp6IsEnabled() and otThreadSetEnabled() ? is it ?

**Reply by jwhui** [reply]
As mentioned in https://github.com/openthread/openthread/discussions/7607#discussioncomment-2598350, OpenThread will automatically handle storing the Active Operational Dataset into non-volatile memory after successfully attaching to a network. You can follow the [Operational Dataset Quick Start Guide](https://github.com/openthread/openthread/blob/main/src/cli/README_DATASET.md#quick-start), to test it out using the example CLI app. You do not need to call `otDatasetGetActive()` and `otDatasetSetActive()`.

**Reply by mithunGhosh13** [reply]
ok,Let me check once.

**Reply by mithunGhosh13** [reply]
Very 1st time while creating otinstance  using otInstanceInitSingle()  is that otinstace needs to be store in flash? 
 The return value of  calling of otIp6IsEnabled() and otThreadSetEnabled() gets false due to otinstace gets zero after reste.
 actually after paired 1st time if power goes down and again power up then controller is getting reset and the instace gets zero.

**Reply by mithunGhosh13** [reply]
How can i retrieve the dataset?

**Reply by jwhui** [reply]
`otInstance` points to the volatile memory used by OpenThread. `otInstanceInitSingle()` effectively calls the constructors on the `otInstance` object. You should not store the `otInstance` object in non-volatile memory.

I suggest taking a look at the CLI example app in [examples/apps/cli/main.c](https://github.com/openthread/openthread/blob/main/examples/apps/cli/main.c).

**Reply by jwhui** [reply]
As mentioned, the device will automatically retrieve the Active Operational Dataset from its parent and store in non-volatile after successfully attaching. If you want to view the Active Operational Dataset saved to non-volatile memory, you can use the `otDatasetGetActive()`.

**Reply by mithunGhosh13** [reply]
yes, i have tried to check the stored data using otDatasetGetActive() in my code . But execution has stuck at the point where i used otDatasetGetActive() function.

**Reply by mithunGhosh13** [reply]
First thing is i am not using the CLI example code. I am using coap_client and coap_server example code which provided by nordic .
I can able to execute the CLI command with the coap example code. i can see devices are not communicating after execute CLI command mention steps in Operational Dataset Quick Start Guide. what are the next steps needs to follow?
![CLI](https://user-images.githubusercontent.com/104003894/164393355-2e0fd9ad-c8c0-44ee-82d9-c877e883eaca.JPG)

**Reply by jwhui** [reply]
Given that this appears to be an issue with Nordic's example code, please reach out via [Nordic DevZone](https://devzone.nordicsemi.com/) for support.

FWIW, I am not able to reproduce your issue using the OpenThread CLI example.
